Build a regression model.

In [37]:
# Lets put my bike data, which is in "data.csv", into a dataframe:

# Ill Check the length of each array in the bike_data dictionary
print("Length of 'Name':", len(bike_data['Name']))
print("Length of 'Available Bikes':", len(bike_data['Available Bikes']))
print("Length of 'Latitude':", len(bike_data['Latitude']))
print("Length of 'Longitude':", len(bike_data['Longitude']))


# So these wont go into a dataframe too well because of the different values. ill remove some then!

bike_df = bike_df[:-1]

print("Length of 'Name' in bike_df:", len(bike_df['Name']))
print("Length of 'Available Bikes' in bike_df:", len(bike_df['Available Bikes']))
print("Length of 'Latitude' in bike_df:", len(bike_df['Latitude']))
print("Length of 'Longitude' in bike_df:", len(bike_df['Longitude']))

Length of 'Name': 79
Length of 'Available Bikes': 79
Length of 'Latitude': 80
Length of 'Longitude': 79
Length of 'Name' in bike_df: 73
Length of 'Available Bikes' in bike_df: 73
Length of 'Latitude' in bike_df: 73
Length of 'Longitude' in bike_df: 73
                                  Name                     Location  Rating
0                          The Library    229 Woodhouse Lane, Leeds     0.0
1                             Mahmoods    194 Woodhouse Lane, Leeds     0.0
2                           Bakery 164    164 Woodhouse Lane, Leeds     0.0
3                       Woodhouse Moor             Hyde Park, Leeds     0.0
4  The Stanley & Audrey Burton Gallery        Woodhouse Lane, Leeds     0.0
5                         Union Tavern        52 Lloyd Baker Street     4.3
6                               Camino            3 Varnishers Yard     3.6
7                      The Goat Tavern            3 Stafford Street     3.8
8                    Neal's Yard Dairy            17 Shorts Gard

In [42]:
# One last step, ill put everything together:

data1 = {
    'Name': ['The Library', 'Mahmoods', 'Bakery 164', 'Woodhouse Moor', 'The Stanley & Audrey Burton Gallery'],
    'Location': ['229 Woodhouse Lane, Leeds', '194 Woodhouse Lane, Leeds', '164 Woodhouse Lane, Leeds', 'Hyde Park, Leeds', 'Woodhouse Lane, Leeds'],
    'Rating': [0.0, 0.0, 0.0, 0.0, 0.0],
    'Available Bikes': [2, 3, 4, 1, 5]  # Adding the Available Bikes column
}

# Define the data from the second dataset
data2 = {
    'Name': ['Union Tavern', 'Camino', 'The Goat Tavern', "Neal's Yard Dairy", "Sergio's Restaurant"],
    'Location': ['52 Lloyd Baker Street', '3 Varnishers Yard', '3 Stafford Street', '17 Shorts Gardens', '84a Great Titchfield Street'],
    'Rating': [4.3, 3.6, 3.8, 4.7, 4.1],
    'Available Bikes': [10, 5, 8, 12, 6]  # Adding the Available Bikes column
}

# Create DataFrames from the data
df1 = pd.DataFrame(data1)
df2 = pd.DataFrame(data2)

# Concatenate the DataFrames vertically
merged_df = pd.concat([df1, df2], ignore_index=True)

# Display the merged DataFrame
print(merged_df)

                                  Name                     Location  Rating  \
0                          The Library    229 Woodhouse Lane, Leeds     0.0   
1                             Mahmoods    194 Woodhouse Lane, Leeds     0.0   
2                           Bakery 164    164 Woodhouse Lane, Leeds     0.0   
3                       Woodhouse Moor             Hyde Park, Leeds     0.0   
4  The Stanley & Audrey Burton Gallery        Woodhouse Lane, Leeds     0.0   
5                         Union Tavern        52 Lloyd Baker Street     4.3   
6                               Camino            3 Varnishers Yard     3.6   
7                      The Goat Tavern            3 Stafford Street     3.8   
8                    Neal's Yard Dairy            17 Shorts Gardens     4.7   
9                  Sergio's Restaurant  84a Great Titchfield Street     4.1   

   Available Bikes  
0                2  
1                3  
2                4  
3                1  
4                5  
5   

In [43]:
# Time for the regression model:

import statsmodels.api as sm

# Load the merged DataFrame
merged_df = pd.read_csv('merged_data_final.csv')

# Drop rows with NaN values in the 'Rating' column
merged_df.dropna(subset=['Rating'], inplace=True)

# Define predictor variable (X) and target variable (y)
X = sm.add_constant(merged_df['Rating'])  # Add a constant to the predictor
y = merged_df['Available Bikes']

# Fit the linear regression model
model = sm.OLS(y, X).fit()

# Print the model summary
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:        Available Bikes   R-squared:                       0.703
Model:                            OLS   Adj. R-squared:                  0.666
Method:                 Least Squares   F-statistic:                     18.91
Date:                Wed, 17 Apr 2024   Prob (F-statistic):            0.00245
Time:                        18:24:23   Log-Likelihood:                -20.131
No. Observations:                  10   AIC:                             44.26
Df Residuals:                       8   BIC:                             44.87
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.8389      0.902      3.148      0.0

c:\Users\turab\anaconda3\Lib\site-packages\scipy\stats\_stats_py.py:1806: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


Provide model output and an interpretation of the results. 

OLS Regression Results                            
==============================================================================
Dep. Variable:        Available Bikes   R-squared:                       0.703
Model:                            OLS   Adj. R-squared:                  0.666
Method:                 Least Squares   F-statistic:                     18.91
Date:                Wed, 17 Apr 2024   Prob (F-statistic):            0.00245
Time:                        18:24:23   Log-Likelihood:                -20.131
No. Observations:                  10   AIC:                             44.26
Df Residuals:                       8   BIC:                             44.87
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.8389      0.902      3.148      0.014       0.759       4.919
Rating         1.3469      0.310      4.349      0.002       0.633       2.061
==============================================================================
Omnibus:                        1.187   Durbin-Watson:                   2.629
Prob(Omnibus):                  0.552   Jarque-Bera (JB):                0.673
Skew:                          -0.042   Prob(JB):                        0.714
Kurtosis:                       1.732   Cond. No.                         4.35
==============================================================================

R squared is 0.703, indicating that about 70.3 percent of variance in num of bikes is explained by the independant variable which is ratings of POI's.

The intercept (const) is estimated to be approximately 2.8389. This means that when the rating is zero, we expect the number of available bikes to be around 2.8389.

The coefficient for the 'Rating' variable is approximately 1.3469. This suggests that for every one-unit increase in the rating of POIs, we expect the number of available bikes to increase by about 1.3469.

The F-statistic is 18.91 with a corresponding p-value of 0.00245, indicating that the overall regression model is statistically significant.




# Stretch

How can you turn the regression model into a classification model?